### ISSUE TITLE
"failed to create vm" when using a user with zsh login shell from sudo -i

### ISSUE THREAD

**Issue by dawidpotocki** [nginx/njs#581]
OS: Arch Linux
njs: 0.7.7

TL;DR njs prints "failed to create vm" and exits if we log into some
user using `sudo -i` that has `zsh` as its login shell and we did not
`unset SHELL`.

---

I found this issue by going to my server, typing `sudo -i` to log in to
root and trying `nginx -t`.

```
2022/09/29 07:23:32 [emerg] 11740#11740: failed to create js VM
```

I thought that something was wrong with my config or something, but no.
It was fine. Was able to restart nginx and it wasn't complaining in logs.
I tried running `njs` by itself and got this message:

```
failed to create vm
```

I exited from my root user and tried running `njs`, it worked. Well
that's odd. Tried `sudo -s` and it… worked too. `sudo su`, yep, worked.

```
interactive njs 0.7.7

v.<Tab> -> the properties and prototype methods of v.

>>
```

Okay, what if I use a different shell? I did `chsh -s /bin/bash` on my
root user, did `sudo -i` and… it also works. Every other shell works
too.

I tried running zsh without any config but that didn't help.

What if I unset some environmental variables?
Okay, so I unsetted pretty much everything and njs started working
after… unsetting `$SHELL`. The weird thing about it is that when I
`export SHELL=/bin/zsh` to put the variable again in the environment,
njs would still work. What if we export the variable without first
unsetting it? It… doesn't work.

What if we run zsh from bash from zsh? It works.

```
$ zsh -c 'bash -c "zsh -c njs"'
```

What if we run zsh from zsh from zsh? It doesn't work.

```
$ zsh -c 'zsh -c "zsh -c njs"'
```

What if we `sudo -i` into a different user than root? It also does not
work as long as the user also had zsh as its login shell.

What if we put njs inside gdb? Well… it works. What if we put zsh that
is executing njs into gdb? Again… works.

Tried it on my laptop too, was able to reproduce this bug.

---

I dug into the source code, but I don't really know C that well and
never worked on njs. Here is what I found:

String "failed to create vm" only shows up in `njs_create_vm`, so it
fails because of `njs_vm_create`'s return value (you guys are very good
at naming stuff).

```c
static njs_vm_t *
njs_create_vm(njs_opts_t *opts, njs_vm_opt_t *vm_options)
{
    u_char      *p, *start;
    njs_vm_t    *vm;
    njs_int_t   ret;
    njs_str_t   path;
    njs_uint_t  i;

    vm = njs_vm_create(vm_options);
    if (vm == NULL) {
        njs_stderror("failed to create vm\n");
        return NULL;
    }
```

In `njs_vm_create` I found that it's gonna return NULL in the following
code.

```c
    vm->options = *options;

    if (options->shared != NULL) {
        vm->shared = options->shared;

    } else {
        ret = njs_builtin_objects_create(vm);
        if (njs_slow_path(ret != NJS_OK)) {
            return NULL;
        }
    }
```

By checking `njs_builtin_objects_create`, I found that this part is
returning `NJS_ERROR`.

```c
    ret = njs_env_hash_init(vm, &shared->env_hash, environ);
    if (njs_slow_path(ret != NJS_OK)) {
        return NJS_ERROR;
    }
```

Then I checked `njs_env_hash_init` and found that this part is returning
`NJS_ERROR`.

```c
        ret = njs_lvlhsh_insert(hash, &lhq);
        if (njs_slow_path(ret != NJS_OK)) {
            njs_internal_error(vm, "lvlhsh insert failed");
            return NJS_ERROR;
        }
```

I wasn't able to determine which thing went wrong in
`njs_lvlhsh_insert`, so… yeah.

---

![screenshot with commands](https://user-images.githubusercontent.com/38681822/192975824-f6af3310-4276-4801-8d94-04702faf8f88.png)
![screenshot with commands](https://user-images.githubusercontent.com/38681822/192975856-d8ec1ed5-0d06-432e-8ff8-b3ca224642c7.png)

**Comment by xeioex** [comment]
Hi @dawidpotocki,

Thanks for the report and the preliminary investigation, it helps.

njs_env_hash_init() is provided with `environ` variable (`man environ`). So the problem seems to be related to environment variables. Can you, please, report the output of `env` in the same environment where nginx/njs is executed?
My guess, the nginx/njs is somehow provided with identical entries like 'FOO=a FOO=b', njs expect them to be always unique, that is the most probable reason why njs_lvlhsh_insert() returns an error if a key is already present.

**Comment by xeioex** [comment]
Hi @dawidpotocki,

Feel free to test the following patch:
```diff
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1664509605 25200
#      Thu Sep 29 20:46:45 2022 -0700
# Node ID d131724367da4389cef3cdbbddc6509c5ff90132
# Parent  b61a7a4f286ecc66ac11bb694dd1d081628d6502
Fixed parsing of environ variables when duplicate keys are present.

This closes #581 issue on Github.

diff --git a/src/njs_builtin.c b/src/njs_builtin.c
--- a/src/njs_builtin.c
+++ b/src/njs_builtin.c
@@ -1835,7 +1835,11 @@ njs_env_hash_init(njs_vm_t *vm, njs_lvlh
         lhq.key_hash = njs_djb_hash(lhq.key.start, lhq.key.length);
 
         ret = njs_lvlhsh_insert(hash, &lhq);
-        if (njs_slow_path(ret != NJS_OK)) {
+        if (njs_slow_path(ret == NJS_ERROR)) {
+            /*
+             * Always using the first element among the duplicates
+             * and ignoring the rest.
+             */
             njs_internal_error(vm, "lvlhsh insert failed");
             return NJS_ERROR;
         }
# HG changeset patch
# User Dmitry Volyntsev <xeioex@nginx.com>
# Date 1664502489 25200
#      Thu Sep 29 18:48:09 2022 -0700
# Node ID 73f668d91d18364a9a9f2a08db6838b03690f906
# Parent  d131724367da4389cef3cdbbddc6509c5ff90132
Improved population of process.env object.

1) Keys are always casted to upper case.
2) Keys and values are converted to safe Unicode strings.

diff --git a/src/njs_builtin.c b/src/njs_builtin.c
--- a/src/njs_builtin.c
+++ b/src/njs_builtin.c
@@ -107,9 +107,6 @@ static const njs_object_type_init_t *con
 };
 
 
-extern char  **environ;
-
-
 njs_inline njs_int_t
 njs_object_hash_init(njs_vm_t *vm, njs_lvlhsh_t *hash,
     const njs_object_init_t *init)
@@ -1793,9 +1790,13 @@ static njs_int_t
 njs_env_hash_init(njs_vm_t *vm, njs_lvlhsh_t *hash, char **environment)
 {
     char                **ep;
-    u_char              *val, *entry;
+    u_char              *dst;
+    ssize_t             length;
+    uint32_t            cp;
     njs_int_t           ret;
+    const u_char        *val, *entry, *s, *end;
     njs_object_prop_t   *prop;
+    njs_string_prop_t   string;
     njs_lvlhsh_query_t  lhq;
 
     lhq.replace = 0;
@@ -1818,14 +1819,28 @@ njs_env_hash_init(njs_vm_t *vm, njs_lvlh
             continue;
         }
 
-        ret = njs_string_set(vm, &prop->name, entry, val - entry);
+        ret = njs_string_create(vm, &prop->name, (char *) entry, val - entry);
         if (njs_slow_path(ret != NJS_OK)) {
             return NJS_ERROR;
         }
 
+        (void) njs_string_prop(&string, &prop->name);
+
+        length = string.length;
+        s = string.start;
+        end = s + string.size;
+        dst = (u_char *) s;
+
+        while (length != 0) {
+            cp = njs_utf8_upper_case(&s, end);
+            dst = njs_utf8_encode(dst, cp);
+            length--;
+        }
+
         val++;
 
-        ret = njs_string_set(vm, &prop->value, val, njs_strlen(val));
+        ret = njs_string_create(vm, &prop->value, (char *) val,
+                                njs_strlen(val));
         if (njs_slow_path(ret != NJS_OK)) {
             return NJS_ERROR;
         }
diff --git a/src/njs_unix.h b/src/njs_unix.h
--- a/src/njs_unix.h
+++ b/src/njs_unix.h
@@ -47,6 +47,8 @@
 
 #include <unistd.h>
 
+extern char  **environ;
+
 #if defined(PATH_MAX)
 #define NJS_MAX_PATH             PATH_MAX
 #else
diff --git a/src/test/njs_unit_test.c b/src/test/njs_unit_test.c
--- a/src/test/njs_unit_test.c
+++ b/src/test/njs_unit_test.c
@@ -13877,6 +13877,9 @@ static njs_unit_test_t  njs_test[] =
     { njs_str("Object.values(process)"),
       njs_str("") },
 
+    { njs_str("Object.keys(process.env).sort()"),
+      njs_str("DUP,TZ") },
+
     { njs_str("Object.values()"),
       njs_str("TypeError: cannot convert undefined argument to object") },
 
@@ -24160,9 +24163,14 @@ main(int argc, char **argv)
         return (ret == NJS_DONE) ? EXIT_SUCCESS: EXIT_FAILURE;
     }
 
+    environ = NULL;
+
     (void) putenv((char *) "TZ=UTC");
     tzset();
 
+    (void) putenv((char *) "DUP=bar");
+    (void) putenv((char *) "dup=foo");
+
     njs_mm_denormals(1);
 
     njs_memzero(&stat, sizeof(njs_stat_t));
```

**Comment by dawidpotocki** [comment]
Huh, you are right, `SHELL` was duplicated.

Looked at that `env` output for so long and didn't notice it. Didn't
even know that such thing would be possible. But I guess it makes sense
why it failed now.

I was now trying to write some patch myself, but I see that you were
unsurprisingly faster than me. But yes, it now works with your patch.
Thanks!

```
root@violet in ~
$ env | grep SHELL
SHELL=/bin/zsh
SHELL=/bin/zsh
root@violet in ~
$ njs
interactive njs 0.7.8

v.<Tab> -> the properties and prototype methods of v.

>> ^C
```

**Comment by dawidpotocki** [comment]
Also just a small note, I noticed that with

```
test=1
test=2
```

dash and bash will output `2` from `echo $test` while zsh will `1`.

Only considering the first variable could potentially cause issues
later, maybe some warning should be provided if they are different?

**Comment by xeioex** [comment]
@dawidpotocki, 

>Didn't even know that such thing would be possible.

Honestly, that was a surprise for me too. It seems, nowadays it is mostly enforced by the apps/libs themselves, whereas low-level `environ` API is just a list of strings, so no uniqueness is enforced.


>EDIT: btw, do you think that I should maybe report this to sudo/zsh too?
>dash and bash will output 2 from echo $test while zsh will 1.

Subclause 7.22.4.6 of the C Standard says: `The set of environment names and the method for altering the environment list are implementation-defined`, so we cannot blame them directly.

Whereas, it is [recommended](https://wiki.sei.cmu.edu/confluence/display/c/ENV02-C.+Beware+of+multiple+environment+variables+with+the+same+effective+name) to consistently choosing a value for duplicate keys.

>Only considering the first variable could potentially cause issues
later, maybe some warning should be provided if they are different?

Agree, it is a good idea, will do.
